Patch-Source: https://gitlab.postmarketos.org/postmarketOS/hkdm/-/merge_requests/8
---
From 6525a23ecf6e7260929842fa35d6aba27f840073 Mon Sep 17 00:00:00 2001
From: Jens Reidel <adrian@travitia.xyz>
Date: Wed, 13 Aug 2025 16:22:01 +0200
Subject: [PATCH] Stop polling removed input devices

This would otherwise result in an infinite busy-loop due to poll
returning immediately.

Signed-off-by: Jens Reidel <adrian@travitia.xyz>
Part-of: https://gitlab.postmarketos.org/postmarketOS/hkdm/-/merge_requests/8
---
 src/main.rs | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/main.rs b/src/main.rs
index 1d3d4b2..79efcb6 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -264,6 +264,24 @@ fn event_loop(m: &mut Hkdm) {
         if poll(&mut fds_arr, 5000).is_err() {
             continue; // No events to read so loop again
         }
+
+        // If revents contain POLLHUP, the input device doesn't exist anymore
+        // Remove all fds that don't exist anymore
+        let mut idx = 0;
+        while idx < m.devices.len() {
+            let poll_fd = &fds_arr[idx];
+
+            if poll_fd
+                .revents()
+                .map_or(false, |revents| revents.contains(PollFlags::POLLHUP))
+            {
+                m.devices.remove(idx);
+                fds_arr.remove(idx);
+            } else {
+                idx += 1;
+            }
+        }
+
         for i in 0..m.devices.len() {
             let device = &mut m.devices[i];
             // From docs, we might need a smart implementation here at some point.
-- 
GitLab

